### PR TITLE
Small things here and there

### PR DESCRIPTION
**Pull request by xloem**
I looked at your code for the first time and noticed these things. I've likely made a mistake updating them, as I'm adding them from a web browser. What do you think of these changes?

- Added path for appending to the state via warmup context
- Updated the state when processing the prompt; it looks like this was missing
- Updated the persistent state when completing generation; it looks like this was missing, but I defaulted to not update it to retain the existing behavior

EDIT:
Other possible improvements that occurred to me as useful:
- passing the state to the callback and returning it from generation
- terminating generation if the callback returns e.g. False
- separation of state from the model object, so multiple contexts could be processed at once by the same model

### CODE REVIEW

File: prwkv/rwkvrnnmodel.py [hunk 1]
```diff
@@ -207,7 +208,7 @@ def generate(self,
             if len(context) > 0:
                 for i in range(len(context)):
                     next_token = context[i:i+1]
-                    _, new_state = self.model.forward(next_token, state)
+                    new_state = self.model.forward(next_token, state, preprocess_only=True)
```
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:211]: yeah makes sense

File: prwkv/rwkvrnnmodel.py [hunk 2]
```diff
@@ -191,7 +191,8 @@ def generate(self,
                  top_p=.9,
                  top_k=20,
                  stop_on_eos=False,
-                 repetition_penalty=2.5)->Union[List[int],None]:
+                 repetition_penalty=2.5,
```
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:194]: cosmedic can you setup a linter to address issues like these?

File: prwkv/rwkvrnnmodel.py [hunk 3]
```diff
@@ -216,6 +217,8 @@ def generate(self,
                        
                     if streaming_callback != None:
                         streaming_callback(next_token[0])
+
```
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:220]: oh damn was this a programming error on my part?

**Comment by ArEnSc** [comment]
btw pm me on the discord if you want to help I have a list of items that would make quite a bit of impact but thank you so far I will look into this a bit more before merging

**Comment by ArEnSc** [comment]
Yeah I am thinking this won't fit into a hugging face api cleanly we might have to be a bit creative to make things easier

**Comment by ArEnSc** [comment]
> EDIT: Other possible improvements that occurred to me as useful:
> 
> * passing the state to the callback and returning it from generation
> * terminating generation if the callback returns e.g. False
> * separation of state from the model object, so multiple contexts could be processed at once by the same model

Yeah I am thinking this won't fit into a hugging face api cleanly we might have to be a bit creative to make things easier

File: prwkv/rwkvrnnmodel.py [hunk 4]
```diff
@@ -83,8 +83,8 @@ def clear_memory(self):
         self.init_state = None
         self.init_logits = None 
         
-    def warmup_with_context(self,context:List[int]):
-        init_state = None 
+    def warmup_with_context(self,context:List[int],clear=True):
```
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:86]: is this meant to allow to build up the context further?

File: prwkv/rwkvrnnmodel.py [hunk 5]
```diff
@@ -262,6 +265,10 @@ def generate(self,
 
                 next_token = token_id
                 
+            if update_state:
```
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:268]: Not sure why we would have this here in the api, for the generate use case, we want it to just generate using what ever initialized state simplifying the work flow. Lmk what you think

**Comment by ArEnSc** [comment]
hey thanks for catching that state bug otherwise!

**Comment by ArEnSc** [comment]
hey I updated the state update bug that was missing thanks for that for the rest we should get some clearity with the intent on the api surface. Thanks @xloem

File: prwkv/rwkvrnnmodel.py (hunk 5 above)
Comment (ArEnSc) [inline prwkv/rwkvrnnmodel.py:268]: I understand what you mean by this ok hmmmm probably should have it as a method on the model rather than thrown into the generate function as a parameter

File: prwkv/rwkvrnnmodel.py (hunk 2 above)
Comment (xloem) [reply prwkv/rwkvrnnmodel.py:194]: this was just to add the update_state parameter on the next line

File: prwkv/rwkvrnnmodel.py (hunk 4 above)
Comment (xloem) [reply prwkv/rwkvrnnmodel.py:86]: yeah; when the setup is all together you can process context with the GPT form and then generate with the RNN form. but maybe it is not the interface you prefer.

**Comment by xloem** [comment]
> Yeah I am thinking this won't fit into a hugging face api cleanly we might have to be a bit creative to make things easier

my preference was to implement the GPT form and transfer the state as a hidden state like with T5. Simplifies to RNN if only one token is passed. The combined GPT-RNN form is kind of complex. I did it once for rwkv 2 or so and i had to really step through it and compare numerical stability of the operations, and do the algebra out by hand. So your approach here makes sense to start with.

> hey I updated the state update bug that was missing thanks for that for the rest we should get some clearity with the intent on the api surface. Thanks @xloem

great.

**Comment by ArEnSc** [comment]
changes added thanks gonna test it later

**Comment by ArEnSc** [comment]
added suggestions with tests
